src/app/services/auth.service.ts

```diff
@@ -0,0 +1,72 @@
+import {Injectable} from '@angular/core';
+import {Router} from '@angular/router';
+import {auth} from 'firebase/app';
+import {Observable} from 'rxjs';
+import {User} from '../models/user.model';
+import * as firebase from 'firebase';
+import {NotificationService} from './notification.service';
+import {of} from 'rxjs/internal/observable/of';
+import {environment} from '../../environments/environment';
+import {AngularFireAuth} from '@angular/fire/auth';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class AuthService {
+  user$: Observable<User | null | undefined> = of(undefined);
+
+  constructor(
+    private router: Router,
+    private notificationService: NotificationService,
+    public angularFireAuth: AngularFireAuth
+  ) {
+    if (environment.requireAuth) {
+        angularFireAuth.onAuthStateChanged(firebaseUser => {
+          if (firebaseUser != null) {
+            const user: User = {uid: firebaseUser.uid, displayName: firebaseUser.displayName};
+            this.user$ = of(user);
+          } else {
+            this.user$ = of(null);
+          }
+        });
+      }
+  }
+
+  googleSignin(): void {
+    const provider = new auth.GoogleAuthProvider();
+    this.angularFireAuth.signInWithPopup(provider).then(result => {
+      const user = result.user;
+      if (user == null) {
+        throw new Error('User is null');
```

Comment:
Under what circumstances will this condition happen in practice? And what is the user-visible behaviour when you throw an `Error` like this?

Reply:
Not sure I got an answer to this one? Apologies if we discussed it and I forgot, but it would be good to note the answer here for the record.

Reply:
We did discuss this, I implemented it this way because I wasn't aware I could write `user?.displayName`. I've changed the implementation now so it doesn't throw an error here.